### DISCUSSION TITLE
Malformed Device ID under Windows 10

### DISCUSSION THREAD

**Discussion by tehn** [hathach/tinyusb#1815] [Q&A]
Hello,

Using example device `cdc_dual_ports` on a standard pico board (rp2040) the PNP Device ID (viewable in System Information) appears to be malformed:

```
USB\VID_CAFE&PID_4002&MI_00\7&2E17A375&0&0000
````

I was expecting to see something similar to that produced by a CDC device using the STM32 HAL, which is something like:

```
USB\VID_CAFE&PID_4002\123456
```

where `123456` is the serial number entered in `usb_descriptors.c`

Currently our application checks the Device ID key to extract the serial number. Is this is bug with device string reporting? As far as I can tell, the device string is never sent (wireshark screenshot attached) and there are also some error packets. I've confirmed this on different hardware and cables.

Any guidance would be greatly appreciated.
<img width="956" alt="Screenshot_20221223_090201" src="https://user-images.githubusercontent.com/246896/209390090-96f28e84-c314-4ce7-a1b5-509dd6d88a32.png">

<img width="876" alt="Screenshot_20221223_090010" src="https://user-images.githubusercontent.com/246896/209390032-8731a90d-652c-497d-b522-4837f7116cad.png">

**Comment by tehn** [comment]
I suppose an alternative version of this question is:

with the example CDC firmware, how is the serial number read within Windows 10?

I currently cannot locate anywhere it is reported.

Thanks!

**Reply by hathach** [reply]
serial is a string descriptor, and is return by callback
https://github.com/hathach/tinyusb/blob/master/examples/device/cdc_dual_ports/src/usb_descriptors.c#L219

you can modify the example to return actual unique ID

**Reply by tehn** [reply]
Apologies, I should have added: the serial number is reported perfectly well by MacOS and Linux. It's only Windows that looks confusing.

I've already worked with the function you reference (in order to send a serial number corresponding to the unique chip ID), but I was seeing this trouble with Windows, so I reverted to working on a known example (the CDC example).

I'm perplexed how the Device ID `USB\VID_CAFE&PID_4002&MI_00\7&2E17A375&0&0000` would be intentional? Is the data after the VID/PID what's expected?

**Reply by HiFiPhile** [reply]
I think it's because you have a composite device, while HAL only has simple device example.

Take this example:
```
        +++++++++++++++++ Device Information ++++++++++++++++++
Device Description       : USB Composite Device
Device Path              : \\?\USB#VID_CAFE&PID_0821#00150011043BD93600000509#{a5dcbf10-6530-11d2-901f-00c04fb951ed} (GUID_DEVINTERFACE_USB_DEVICE)
Kernel Name              : \Device\USBPDO-10
Device ID                : USB\VID_CAFE&PID_0821\00150011043BD93600000509
Hardware IDs             : USB\VID_CAFE&PID_0821&REV_0100 USB\VID_CAFE&PID_0821
Driver KeyName           : {36fc9e60-c465-11cf-8056-444553540000}\0076 (GUID_DEVCLASS_USB)
Driver                   : \SystemRoot\System32\drivers\usbccgp.sys (Version: 10.0.19041.1949  Date: 2022-09-23)
Driver Inf               : C:\Windows\inf\usb.inf
Legacy BusType           : PNPBus
Class                    : USB
Class GUID               : {36fc9e60-c465-11cf-8056-444553540000} (GUID_DEVCLASS_USB)
Service                  : usbccgp
Enumerator               : USB
Location Info            : Port_#0001.Hub_#0006
Location IDs             : PCIROOT(0)#PCI(0102)#PCI(0000)#PCI(0800)#PCI(0001)#USBROOT(0)#USB(6)#USB(1), ACPI(_SB_)#ACPI(PCI0)#ACPI(BXBR)#ACPI(BYUP)#ACPI(BYD8)#ACPI(XHC1)#ACPI(RHUB)#ACPI(PRT6)#USB(1)
Container ID             : {c6f6bc07-2b6c-5da7-b9be-b159d562717d}
Manufacturer Info        : (Standard USB Host Controller)
Capabilities             : 0x94 (Removable, UniqueID, SurpriseRemovalOK)
Status                   : 0x0180600A (DN_DRIVER_LOADED, DN_STARTED, DN_DISABLEABLE, DN_REMOVABLE, DN_NT_ENUMERATOR, DN_NT_DRIVER)
Problem Code             : 0
Address                  : 1
HcDisableSelectiveSuspend: 0
EnableSelectiveSuspend   : 0
SelectiveSuspendEnabled  : 0
EnhancedPowerMgmtEnabled : 0
IdleInWorkingState       : 0
WakeFromSleepState       : 0
Power State              : D0 (supported: D0, D2, D3, wake from D0, wake from D2)
 Child Device 1          : PWM_Ctrl V2 (WinUsb Device)
  Device Path 1          : \\?\USB#VID_CAFE&PID_0821&MI_00#9&1fedf0d4&0&0000#{dee824ef-729b-4a0e-9c14-b7117d33a817} (GUID_DEVINTERFACE_WINUSB)
  Device Path 2          : \\?\USB#VID_CAFE&PID_0821&MI_00#9&1fedf0d4&0&0000#{98762270-47cd-44cc-9272-8f40c8ce815f}
  Kernel Name            : \Device\00000076
  Device ID              : USB\VID_CAFE&PID_0821&MI_00\9&1FEDF0D4&0&0000
  Class                  : USBDevice
  Driver KeyName         : {88bae032-5a81-49f0-bc3d-a4ff138216d6}\0012 (GUID_DEVCLASS_USBDEVICE)
  Service                : WINUSB
  Location               : 0009.0000.0001.006.001.000.000.000.000
  LocationPaths          : PCIROOT(0)#PCI(0102)#PCI(0000)#PCI(0800)#PCI(0001)#USBROOT(0)#USB(6)#USB(1)#USBMI(0)  PCIROOT(0)#PCI(0102)#PCI(0000)#PCI(0800)#PCI(0001)#USBROOT(0)#USB(6)#USB(1)#USB(1)  ACPI(_SB_)#ACPI(PCI0)#ACPI(BXBR)#ACPI(BYUP)#ACPI(BYD8)#ACPI(XHC1)#ACPI(RHUB)#ACPI(PRT6)#USB(1)#USBMI(0)  ACPI(_SB_)#ACPI(PCI0)#ACPI(BXBR)#ACPI(BYUP)#ACPI(BYD8)#ACPI(XHC1)#ACPI(RHUB)#ACPI(PRT6)#USB(1)#USB(1)
 Child Device 2          : DFU PWM_Ctrl (WinUsb Device)
  Device Path 1          : \\?\USB#VID_CAFE&PID_0821&MI_01#9&1fedf0d4&0&0001#{56782270-47cd-44cc-9272-8f40c8ce815f}
  Device Path 2          : \\?\USB#VID_CAFE&PID_0821&MI_01#9&1fedf0d4&0&0001#{dee824ef-729b-4a0e-9c14-b7117d33a817} (GUID_DEVINTERFACE_WINUSB)
  Kernel Name            : \Device\00000077
  Device ID              : USB\VID_CAFE&PID_0821&MI_01\9&1FEDF0D4&0&0001
  Class                  : USBDevice
  Driver KeyName         : {88bae032-5a81-49f0-bc3d-a4ff138216d6}\0013 (GUID_DEVCLASS_USBDEVICE)
  Service                : WINUSB
  Location               : 0009.0000.0001.006.001.000.000.000.000
  LocationPaths          : PCIROOT(0)#PCI(0102)#PCI(0000)#PCI(0800)#PCI(0001)#USBROOT(0)#USB(6)#USB(1)#USBMI(1)  ACPI(_SB_)#ACPI(PCI0)#ACPI(BXBR)#ACPI(BYUP)#ACPI(BYD8)#ACPI(XHC1)#ACPI(RHUB)#ACPI(PRT6)#USB(1)#USBMI(1)
```

Serial number is shown in `Device ID                : USB\VID_CAFE&PID_0821\00150011043BD93600000509` , but in child devices (corresponding to TUSB interfaces) it's not shown `Device ID              : USB\VID_CAFE&PID_0821&MI_00\9&1FEDF0D4&0&0000
`

**Reply by tehn** [reply]
Thank you, that explains the problem I'm seeing.

Am I correct that the TinyUSB library _always_ enumerates as a composite device, even if only a single CDC (or MIDI, etc) device is defined? (This is what my experimentation has shown). Is there no way to have it declare itself as simply a single CDC, not as a composite?

**Reply by tehn** [reply]
Solved my problem by changing to `TUSB_CLASS_CDC`

https://github.com/hathach/tinyusb/blob/master/examples/device/cdc_dual_ports/src/usb_descriptors.c#L155

Thank you all!
